massa-ledger/src/ledger_db.rs

```diff
@@ -0,0 +1,187 @@
+
+/// Ledger sub entry enum
+pub enum LedgerSubEntry {
+    /// Balance
+    Balance,
+    /// Bytecode
+    Bytecode,
+    /// Datastore entry
+    Datastore(Hash),
+}
+
+pub(crate) struct LedgerDB(DB);
+
+// NOTE: still handle separate bytecode for now to avoid too many refactoring at once
+macro_rules! bytecode_key {
+    ($addr:ident) => {
+        format!("b:{}", $addr).as_bytes()
+    };
+}
+
+macro_rules! data_key {
+    ($addr:ident, $key:ident) => {
+        format!("{}:{}", $addr, $key).as_bytes()
+    };
+}
+
+impl LedgerDB {
+    pub fn new() -> Self {
+        // db options
+        let mut db_opts = Options::default();
+        db_opts.create_if_missing(true);
+        db_opts.create_missing_column_families(true);
+
+        // database init
+        let db = DB::open_cf_descriptors(
+            &db_opts,
+            DB_PATH,
+            vec![
+                ColumnFamilyDescriptor::new(LEDGER_CF, Options::default()),
+                ColumnFamilyDescriptor::new(METADATA_CF, Options::default()),
+            ],
+        )
+        .expect(OPEN_ERROR);
+
+        // return database
+        LedgerDB(db)
+    }
+
+    pub fn put(&mut self, addr: &Address, ledger_entry: LedgerEntry) {
+        let mut batch = WriteBatch::default();
+        let key = addr.to_bytes();
+        let handle = self.0.cf_handle(LEDGER_CF).expect(CF_ERROR);
+
+        // balance
+        batch.put_cf(
+            handle,
+            key,
+            ledger_entry.parallel_balance.to_raw().to_be_bytes(),
+        );
+
+        // bytecode
+        batch.put_cf(handle, bytecode_key!(addr), ledger_entry.bytecode);
+
+        // datastore
+        for (hash, entry) in ledger_entry.datastore {
+            batch.put_cf(handle, data_key!(addr, hash), entry);
+        }
+
+        // write batch
+        self.0.write(batch).expect(CRUD_ERROR);
+    }
+
+    pub fn update(&mut self, addr: &Address, entry_update: LedgerEntryUpdate) {
+        let mut batch = WriteBatch::default();
+        let key = addr.to_bytes();
+        let handle = self.0.cf_handle(LEDGER_CF).expect(CF_ERROR);
+
+        // balance
+        if let SetOrKeep::Set(balance) = entry_update.parallel_balance {
+            batch.put_cf(handle, key, balance.to_raw().to_be_bytes());
+        }
+
+        // bytecode
+        if let SetOrKeep::Set(bytecode) = entry_update.bytecode {
+            batch.put_cf(handle, bytecode_key!(addr), bytecode);
+        }
+
+        // datastore
+        for (hash, update) in entry_update.datastore {
+            match update {
+                SetOrDelete::Set(entry) => batch.put_cf(handle, data_key!(addr, hash), entry),
+                SetOrDelete::Delete => batch.delete_cf(handle, data_key!(addr, hash)),
+            }
+        }
+
+        // write batch
+        self.0.write(batch).expect(CRUD_ERROR);
+    }
+
+    pub fn delete(&self, _addr: &Address) {
+        // TODO: define how we want to handle this first
```

Comment:
I think we can fetch all the data of an address using prefix.
Here is the docs of rocksdb about it : https://github.com/facebook/rocksdb/wiki/Iterator
And an issue of rust repo that show some examples : https://github.com/rust-rocksdb/rust-rocksdb/issues/577

Reply:
Prefix was implemented for the intermediate bootstrap, will use it in the delete function.